sql: introduce expect_and_ignore_not_visible_columns_in_copy session var

otan opened the pull request:
Release note (sql change): Introduced a
`expect_and_ignore_not_visible_columns_in_copy` session variable. If this
is set, COPY with no column specifiers will assume hidden columns are in
the copy data, but will ignore them when applying the COPY.

Resolves #79007

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79016)
<!-- Reviewable:end -->

otan commented:
(i think this PR is useful anyway in case, e.g. the user applies a schema migration including a not visible column when the rowid column gets deleted).

rafiss reviewed (changes requested):
just minor questions and nits!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-:-MzSrIWkE71U6IZamSI_:bwgf6gh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/vars.go, line 2030 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSo9kt53ymMHltVYh3:-MzSo9ku7M-1ezLuWd_u:b-1i213q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/vars.go#L2030)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`copy_expect_and_ignore_not_visible_columns`: {
> ```

nit: what do you think about the name `ignore_hidden_columns_in_copy`

- "expect_and_ignore" seems repetitive
- is "hidden" the same as "not_visible" ?
___
*[pkg/sql/pgwire/testdata/pgtest/copy, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSpIyT78JY7PvDgC3D:-MzSpIyT78JY7PvDgC3E:b5xk2ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/pgwire/testdata/pgtest/copy#L67)):*
> ```
> ----
> 
> until crdb_only
> ```

could use `keepErrMessage` to make sure it's what we want (and in below cases too)
___
*[pkg/sql/pgwire/testdata/pgtest/copy, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSr26K45fCo6eDgybK:-MzSr26K45fCo6eDgybL:bpyl6xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/pgwire/testdata/pgtest/copy#L128)):*
> ```
> send crdb_only
> Query {"String": "COPY copy_hidden_table FROM STDIN CSV"}
> CopyData {"Data": "2,blah,good,pk\n"}
> ```

hm i didn't follow why there are 4 columns (i thought it would be 3)
___
*[pkg/sql/pgwire/testdata/pgtest/copy, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSr7Yz65GY0L_vl1oc:-MzSr7Z-5lzzsSqhWC3G:b-42at9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/pgwire/testdata/pgtest/copy#L154)):*
> ```
> 
> send crdb_only
> Query {"String": "SELECT * FROM copy_hidden_table ORDER BY id"}
> ```

should we explicitly SELECT the hidden columns (in a separate test), just to verify that the behavior is well defined?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79016)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/vars.go, line 2030 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSo9kt53ymMHltVYh3:-MzSxdLJ7lULj7gBDBm1:bhlbqzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/vars.go#L2030)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: what do you think about the name `ignore_hidden_columns_in_copy`

- "expect_and_ignore" seems repetitive
- is "hidden" the same as "not_visible" ?
</blockquote></details>

* i wanted "expect" *and* "ignore" because copy now goes out of the way to expect hidden columns to appear, and then to ignore them. just saying `ignore` imo opens the possibility of it not necessarily expecting the hidden columns; but atm it's very difficult to code in that assumption and deal with both cases.
* not_visible is how it is verbalised via SQL (i.e. rowid INT NOT VISIBLE ...)
___
*[pkg/sql/pgwire/testdata/pgtest/copy, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSpIyT78JY7PvDgC3D:-MzSyoxD4XVR4Nb5N24e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/pgwire/testdata/pgtest/copy#L67)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could use `keepErrMessage` to make sure it's what we want (and in below cases too)
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/testdata/pgtest/copy, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSr26K45fCo6eDgybK:-MzSypXPAaKLdvYyT5IW:biwbs04) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/pgwire/testdata/pgtest/copy#L128)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm i didn't follow why there are 4 columns (i thought it would be 3)
</blockquote></details>

ah! added a comment. `rowid`
___
*[pkg/sql/pgwire/testdata/pgtest/copy, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSr7Yz65GY0L_vl1oc:-MzSyrms4o2Q1nApJ_N0:bt2dayz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/pgwire/testdata/pgtest/copy#L154)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should we explicitly SELECT the hidden columns (in a separate test), just to verify that the behavior is well defined?
</blockquote></details>

yeah i'll do it for `nv` but not `rowid` because `rowid` is non-deterministic.


<!-- Sent from Reviewable.io -->

otan commented:
*[pkg/sql/vars.go, line 2030 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSo9kt53ymMHltVYh3:-MzSzH1X3CMHvsUVIkmd:b-4sjjw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/vars.go#L2030)):*
> is "hidden" the same as "not_visible" ?

more concretely, "yes"


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
left an optional bikeshed for you

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-:-MzVYaeLBEfJDDIg7USg:b-sh8lk2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/vars.go, line 2030 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79016#-MzSo9kt53ymMHltVYh3:-MzVXjEC7LufoRymi_6G:b-277275) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1a18afa07247da01ac5ca9164c034baa4d608e/pkg/sql/vars.go#L2030)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> is "hidden" the same as "not_visible" ?

more concretely, "yes"
</blockquote></details>

ah ok; fine on those things. my last bit of bikeshedding is that the `copy_` at the beginning is confusing to me

i keep interpreting it as: "copy, expect, and ignore the not-visible columns"

but it should be interpreted as "while doing COPY, we should expect and ignore not-visible columns"

if this resonates with you, maybe a name like `expect_and_ignore_not_visible_columns_in_copy`


<!-- Sent from Reviewable.io -->

otan committed:
sql: introduce expect_and_ignore_not_visible_columns_in_copy session var

Release note (sql change): Introduced a
`expect_and_ignore_not_visible_columns_in_copy` session variable. If
this is set, COPY with no column specifiers will assume hidden columns
are in the copy data, but will ignore them when applying the COPY.

otan commented:
bors r=rafiss
